cmake/Util.cmake

```diff
@@ -165,21 +165,27 @@ endfunction()
 #
 # Passing CMAKE_BUILD_TYPE for multi-config generators will not only not be
 # used, but also generate a warning for the user.
-function(set_default_buildtype)
+function(set_default_buildtype BUILD_TYPE)
   set(allowableBuildTypes Debug Release MinSizeRel RelWithDebInfo)
+  if(NOT BUILD_TYPE IN_LIST allowableBuildTypes)
+    message(FATAL_ERROR "Invalid build type: ${BUILD_TYPE}")
+  endif()
 
   get_property(isMultiConfig GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
   if(isMultiConfig)
+    # Multi-config generators use the first element in CMAKE_CONFIGURATION_TYPES as the default build type
+    list(INSERT allowableBuildTypes 0 ${BUILD_TYPE})
+    list(REMOVE_DUPLICATES allowableBuildTypes)
     set(CMAKE_CONFIGURATION_TYPES ${allowableBuildTypes} PARENT_SCOPE)
     if(CMAKE_BUILD_TYPE)
       message(WARNING "CMAKE_BUILD_TYPE specified which is ignored on \
-      multi-configuration generators. Defaulting to Debug build type.")
+      multi-configuration generators. Defaulting to ${BUILD_TYPE} build type.")
     endif()
   else()
     set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "${allowableBuildTypes}")
     if(NOT CMAKE_BUILD_TYPE)
-      message(STATUS "CMAKE_BUILD_TYPE not specified, default is 'Debug'")
-      set(CMAKE_BUILD_TYPE Debug CACHE STRING "Choose the type of build" FORCE)
+      message(STATUS "CMAKE_BUILD_TYPE not specified, default is '${BUILD_TYPE}'")
+      set(CMAKE_BUILD_TYPE ${BUILD_TYPE} CACHE STRING "Choose the type of build" FORCE)
     elseif(NOT CMAKE_BUILD_TYPE IN_LIST allowableBuildTypes)
       message(FATAL_ERROR "Invalid build type: ${CMAKE_BUILD_TYPE}")
```

Comment:
Hmm, is this still needed?

Reply:
This check happens if someone sets `CMAKE_BUILD_TYPE` themselves, e.g. `make CMAKE_BUILD_TYPE=Blorf`

Reply:
Wait, you're right. OK thanks for checking.